File: src/math/quaternion.c
#include "quaternion.h"

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

il_quat il_quat_new(float x, float y, float z, float w)
{
    il_quat q;
    q.x = x;
    q.y = y;
    q.z = z;
    q.w = w;
    return q;
}

il_quat il_quat_mul(il_quat a, il_quat b)
{
    il_quat n;
    n.w = a.w * b.w - a.x * b.x - a.y * b.y - a.z * b.z;
    n.x = a.w * b.x + a.x * b.w + a.y * b.z - a.z * b.y;
    n.y = a.w * b.y + a.y * b.w + a.z * b.x - a.x * b.z;
    n.z = a.w * b.z + a.z * b.w + a.x * b.y - a.y * b.x;
    return n;
}

il_quat il_quat_lerp(il_quat a, il_quat b, float t)
{
    il_quat q;
    q.x = a.x * (1-t) + b.x * t;
    q.y = a.y * (1-t) + b.y * t;
    q.z = a.z * (1-t) + b.z * t;
    q.w = a.w * (1-t) + b.w * t;
    return il_quat_normalize(q);
}

il_quat il_quat_fromYPR(float yaw, float pitch, float roll)
{
    il_quat q;
    float c1 = cos(yaw/2);
    float c2 = cos(pitch/2);
    float c3 = cos(roll/2);
    float s1 = sin(yaw/2);
    float s2 = sin(pitch/2);
    float s3 = sin(roll/2);
    q.x = (c1 * c2 * c3) - (s1 * s2 * s3);
    q.y = (s1 * s2 * c3) + (c1 * c2 * s3);
    q.z = (s1 * c2 * c3) + (c1 * s2 * s3);
    q.w = (c1 * s2 * c3) - (s1 * c2 * s3);
    return q;

}

il_quat il_quat_fromAxisAngle(float x, float y, float z, float a)
{
    il_quat q;
    float s = sinf(a/2);
    if (abs(x*x + y*y + z*z - 1) > 0.001) {
        float m = 1/sqrt(x*x+y*y+z*z);
        x *= m;
        y *= m;
        z *= m;
    }
    q.x = s * x;
    q.y = s * y;
    q.z = s * z;
    q.w = cosf(a/2);
    return il_quat_normalize(q);
}

il_quat il_quat_normalize(il_quat a)
{
    il_quat q;
    float len = il_quat_len(a);
    if (len > 0) {
        float ilen = 1/len;
        q.x = a.x * ilen;
        q.y = a.y * ilen;
        q.z = a.z * ilen;
        q.w = a.w * ilen;
    } else {
        q.x = 0;
        q.y = 0;
        q.z = 0;
        q.w = 1;
    }
    return q;
}

il_quat il_quat_invert(il_quat q)
{
    il_quat r;
    r.x = -q.x;
    r.y = -q.y;
    r.z = -q.z;
    r.w = q.w;
    return r;
}

float il_quat_len(il_quat a)
{
    return sqrt(il_quat_dot(a, a));
}

float il_quat_dot(il_quat a, il_quat b)
{
    return a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w;
}

char *il_quat_print(il_quat q, char *buf, unsigned len)
{
    unsigned flen = snprintf(NULL, 0, "[% .2f % .2f % .2f % .2f]", q.x, q.y, q.z, q.w);
    ++flen;
    if (!buf || flen > len) {
        buf = realloc(buf, flen);
    }
    snprintf(buf, flen, "[% .2f % .2f % .2f % .2f]", q.x, q.y, q.z, q.w);
    return buf;
}

File: script/math/quaternion.lua
--- Quaternion type
-- @type quaternion
local ffi = require "ffi"

require "math.scalar_defs"

local vector3-- = require "vector3"

ffi.cdef [[

il_quat il_quat_new(float x, float y, float z, float w);
il_quat il_quat_mul(const il_quat a, const il_quat b);
il_quat il_quat_lerp(const il_quat a, const il_quat b, float t);
il_quat il_quat_fromYPR(float yaw, float pitch, float roll);
il_quat il_quat_fromAxisAngle(float x, float y, float z, float a);
il_quat il_quat_normalize(const il_quat a);
float il_quat_len(const il_quat a);
float il_quat_dot(const il_quat a, const il_quat b);

]]

local quaternion = {}

--- FFI CType of quaternions
quaternion.type = ffi.typeof("il_quat");

local axis = {"x", "y", "z", "w"}
function index(t,k)
    vector3 = vector3 or require "math.vector3"
    if k == "x" then
        return t.ptr.x
    elseif k == "y" then
        return t.ptr.y
    elseif k == "z" then
        return t.ptr.z
    elseif k == "w" or k == "i" then
        return t.ptr.w
    elseif k == "v" then
        return vector3(t.x, t.y, t.z)
    elseif k == "len" then
        return il_quat_len(t.ptr)
    elseif k == "dot" then
        return il_quat_dot(t.ptr)
    elseif k == "normal" then
        return quaternion.wrap(il_quat_normalize(t.ptr))
    end
    return quaternion[axis[k]]
end

function newindex(t, k, v)
    vector3 = vector3 or require "math.vector3"
    if k == "x" or k == "y" or k == "z" or k == "w" then
        assert(type(v) == "number")
        t.ptr[k] = v
    elseif k == "v" then
        assert(vector3.check(v))
        t.ptr.x = v.x
        t.ptr.y = v.y
        t.ptr.z = v.z
    elseif k == "i" then
        assert(type(v) == "number")
        t.ptr.w = v
    end
    error("Invalid key \""..tostring(k).."\" in quaternion")
end

function mul(a, b)
    vector3 = vector3 or require "math.vector3"
    if quaternion.check(b) then
        return quaternion.wrap(modules.math.il_quat_mul(a.ptr, b.ptr))
    elseif vector3.check(b) then
        return b * a
    else
        error("Expected quaternion or vector3")
    end
end

function lerp(a, b, t)
    return quaternion.wrap(modules.math.il_quat_lerp(a,b,t))
end

function gc(obj)
    modules.math.il_quat_free(obj.ptr)
end

function ts(self)
    return "["..self.ptr.x..", "..self.ptr.y..", "..self.ptr.z..", "..self.ptr.w.."]"
end

--- Converts a cdata to a quaternion
function quaternion.wrap(ptr)
    local obj = {}
    obj.ptr = ptr;
    obj.T = "quat"
    setmetatable(obj, {__index = index, __newindex=newindex, __mul = mul, __gc = gc, __tostring = ts})
    return obj
end

--- Tests whether an object is a quaternion
function quaternion.check(obj)
    return type(obj) == "table" and obj.T == "quat" and ffi.istype(quaternion.type, obj.ptr)
end

--- Creates a new quaternion
-- Can be passed YPR, Axis (as vec3) Angle, and Axis (as numbers) Angle.
function quaternion.create(...)
    vector3 = vector3 or require "math.vector3"
    local args = {...};
    if #args == 1 then
        if ffi.istype("il_quat", args[1]) then
            return quaternion.wrap(args[1])
        elseif quaternion.check(args[1]) then
            return quaternion.wrap(args[1].ptr)
        else
            assert(false)
        end
    elseif #args == 3 then -- YPR
        assert(type(args[1]) == "number" and type(args[2]) == "number" and type(args[3]) == "number")
        return quaternion.wrap(modules.math.il_quat_fromYPR(args[1], args[2], args[3]))
    elseif #args == 2 then -- axis angle with vec3
        assert(vector3.check(args[1]))
        assert(type(args[2]) == "number")
        return quaternion.wrap(modules.math.il_quat_fromAxisAngle(args[1].x, args[1].y, args[1].z, args[2]))
    elseif #args == 4 then -- raw quaternion
        assert(type(args[1]) == "number" and type(args[2]) == "number" and type(args[3]) == "number" and type(args[4]) == "number")
        --return quaternion.wrap(modules.math.il_quat_fromAxisAngle(args[1], args[2], args[3], args[4], nil))
        return quaternion.wrap(modules.math.il_quat_new(args[1], args[2], args[3], args[4]))
    end
    error "Don't know how to handle args"
end

setmetatable(quaternion, {__call = function(self, ...) return quaternion.create(...) end})

return quaternion;

